### ISSUE TITLE
Received an unexpected EOF or 0 bytes from the transport stream

### ISSUE THREAD

**Issue by mikemeinz** [robinrodricks/FluentFTP#410]
**FTP OS:** Windows

**FTP Server:** Detected FTP server: WindowsServerIIS

**Computer OS:** 
Client: Microsoft Windows 10 Version 1803 (Build 17134.706)
VB .NET with .NET Framework version 4.8
FluentFTP Version 25.0.1

Using **DownloadFile()**, frequently get FTPException InnerException message: "Received an unexpected EOF or 0 bytes from the transport stream. -2146232800"

Immediately retrying the **DownloadFile()** one or more times and the file download will eventually work OK.

Have used a TIMEOUT value ranging from 6000ms to 60000ms with same results.
Have used FtpDataConnectionType.PASV with same results.

Trace log shows "Exception thrown: 'FluentFTP.FtpException' in FluentFTP.dll"at point of error.

**VB .NET code used to connect to the FTP server**
```
                    client = New FtpClient(strServerName, strUsername, strPassword) With {
                        .EncryptionMode = FtpEncryptionMode.Explicit,
                        .SslProtocols = Authentication.SslProtocols.Tls12,
                        .DataConnectionConnectTimeout = TIMEOUT,
                        .DataConnectionReadTimeout = TIMEOUT,
                        .ReadTimeout = TIMEOUT,
                        .ConnectTimeout = TIMEOUT,
                        .InternetProtocolVersions = FtpIpVersion.IPv4,
                        .DownloadRateLimit = 0,
                        .RetryAttempts = 2,
                        .DataConnectionType = FtpDataConnectionType.EPSV,
                        .DownloadDataType = FtpDataType.Binary
                    }
                    AddHandler client.ValidateCertificate, New FtpSslValidation(AddressOf OnValidateCertificate)
                    client.Connect()
```

**VB .NET code used to download a file**
```
                     client.DownloadFile(strLocalFolder & f.Filename, "/" & strFolder & "/" & f.Filename, FtpLocalExists.Overwrite, FtpVerify.Retry, DownloadProgress)
```
Use of **FtpTrace.AddListener(New ConsoleTraceListener())** caused some duplicate trace log entries.
**Trace Log with IP, Username and Password obfuscated**
TraceLog.txt attached
[TraceLog.txt](https://github.com/robinrodricks/FluentFTP/files/3287585/TraceLog.txt)

**Sample program**
```
Imports System.Security
Imports System.Threading
Imports FluentFTP
Module TestFTPS

    Dim client As FtpClient = Nothing
    Dim bOK As Boolean = False
    Dim TryCount As Integer = 0
    Dim TIMEOUT As Integer = 10000
    Const MAXRETRYCOUNT As Integer = 5
    Dim InitialLogin As Boolean = True

    Dim DownloadProgress As Progress(Of FtpProgress) = New Progress(Of FtpProgress)(Sub(x)
                                                                                        If x.Progress > 0 Then
                                                                                            Console.WriteLine(Format(x.Progress / 100, "###.00%"))
                                                                                        End If
                                                                                    End Sub)

    Sub Main()
        client = New FtpClient("xxxxxx", "xxx", "xxxx") With {
                        .EncryptionMode = FtpEncryptionMode.Explicit,
                        .SslProtocols = Authentication.SslProtocols.Tls12,
                        .DataConnectionConnectTimeout = TIMEOUT,
                        .DataConnectionReadTimeout = TIMEOUT,
                        .ReadTimeout = TIMEOUT,
                        .ConnectTimeout = TIMEOUT,
                        .InternetProtocolVersions = FtpIpVersion.IPv4,
                        .DownloadRateLimit = 0,
                        .RetryAttempts = 2,
                        .DataConnectionType = FtpDataConnectionType.EPSV,
                        .DownloadDataType = FtpDataType.Binary
                    }
        AddHandler client.ValidateCertificate, New FtpSslValidation(AddressOf OnValidateCertificate)
        client.Connect()
        Do
            Try
                TryCount += 1
                Console.WriteLine("**** Try #" & TryCount.ToString)
                client.DownloadFile("c:\temp\cntyRaceQuestions.txt", "/20181106/cntyRaceQuestions.txt", FtpLocalExists.Overwrite, FtpVerify.Retry, DownloadProgress)
                bOK = True
            Catch ex As FtpAuthenticationException
                Console.WriteLine("ERROR: " & ex.Message)
                TryCount = MAXRETRYCOUNT + 1
            Catch ex As FtpException
                Dim s As String = ""
                Console.WriteLine("FTP Exception=" & ex.Message)
                If ex.Message.Contains("InnerException") Then
                    If ex.InnerException IsNot Nothing Then
                        s = ex.InnerException.Message
                        Console.WriteLine("Inner Exception=" & ex.InnerException.Message & " " & ex.InnerException.HResult)
                    Else
                        s = ex.Message
                    End If
                Else
                    s = ex.Message
                End If
                Console.WriteLine("ERROR: " & s)
                If client.LastReply.ErrorMessage IsNot Nothing Then Console.WriteLine("ErrorMessage=" & client.LastReply.ErrorMessage)
                If client.LastReply.Message IsNot Nothing Then Console.WriteLine("Message=" & client.LastReply.Message)
                If client.LastReply.InfoMessages IsNot Nothing Then Console.WriteLine("InfoMessage=" & client.LastReply.InfoMessages)
                If client.LastReply.Code.Length > 0 Then Console.WriteLine("Code=" & client.LastReply.Code)
            Catch ex As Exception
                Console.WriteLine("ERROR: " & ex.Message)
            End Try
            If Not bOK Then
                Console.WriteLine("5 second pause")
                Thread.Sleep(5000)
            End If
        Loop Until bOK OrElse TryCount >= MAXRETRYCOUNT
        Try
            client.Disconnect()
        Catch
        End Try
        Console.WriteLine("Done. Press Enter to exit.")
        Console.ReadLine()
    End Sub

    Sub OnValidateCertificate(ByVal control As FtpClient, ByVal e As FtpSslValidationEventArgs)
        If InitialLogin Then
            Dim arCert() As String = e.Certificate.Subject.Split(CType(",", Char()))
            Console.WriteLine("Connected to " & arCert(0).Substring(3)) ' CN=<SERVERNAME>
            InitialLogin = False
        End If
        e.Accept = True
    End Sub

End Module
```

**Output**
```
**** Try #1
Connected to xx.xx.xx.xx
11.83%
23.65%
35.48%
47.31%
59.13%
70.96%
82.78%
94.61%
100.00%
FTP Exception=Error while downloading the file from the server. See InnerException for more info.
Inner Exception=Received an unexpected EOF or 0 bytes from the transport stream. -2146232800
ERROR: Received an unexpected EOF or 0 bytes from the transport stream.
ErrorMessage=
Message=Data connection already open; Transfer starting.
Code=125
5 second pause
**** Try #2
11.83%
23.65%
35.48%
47.31%
59.13%
70.96%
82.78%
94.61%
100.00%
Done. Press Enter to exit.
```

**Comment by mikemeinz** [comment]
This seems to solve the issue. Please consider applying this change to the FluentFTP library. Thanks.

Current FtpClient_HighLevel.cs line 1875 in the **ResumeDownload()** method:
`	if (ex.InnerException != null || ex.Message.StartsWith("Unexpected EOF for remote file")) {`

My program sees this InnerException Message:
	Received an unexpected EOF or 0 bytes from the transport stream.

I changed the line in **ResumeDownload()** to this and tested successfully:
`	if (ex.InnerException != null || ex.Message.ToLower.Contains("unexpected eof")) {`

**Comment by robinrodricks** [comment]
Fix released in 25.0.2. Please check!

https://www.nuget.org/packages/FluentFTP/25.0.2

**Comment by mikemeinz** [comment]
Testing showed no change in behavior.

I found that In **IsKnownError()** the **reply** is converted to all lowercase letters. The **unexpectedEOFStrings** strings to be checked contain both uppercase and lowercase letters. They never match.

`		private static string[] unexpectedEOFStrings = new string[] { "Unexpected EOF for remote file", "Received an unexpected EOF", "Unexpected EOF" };`

```
		public static bool IsKnownError(this string reply, string[] strings) {
			reply = reply.ToLower();
			foreach (string msg in strings) {
				if (reply.Contains(msg)) {
					return true;
				}
			}
			return false;
		}

```

**Comment by robinrodricks** [comment]
Sorry!

https://www.nuget.org/packages/FluentFTP/25.0.3

**Comment by mikemeinz** [comment]
Thanks. Works fine for **DownloadFile()**. 

Still occasionally get _"Received an unexpected EOF or 0 bytes from the transport stream."_ error message when executing **GetListing()**.

**Comment by robinrodricks** [comment]
Can you share logs? See the FAQ on how to get those...

**Comment by mikemeinz** [comment]
Trace Log Line 9883 is where the exception "_Received an unexpected EOF or 0 bytes from the transport stream._" was returned from **GetListing()**.
Immediately following line 9883 is the application program's successful retry of the **GetListing()**.

In the attached trace log, all exceptions returned to the application program during this test are denoted in the log with the prefix **ERROR:**. When the application program receives an exception from FluentFTP, it retries the FluentFTP method. Almost always, it is successful on the retry. If not, the application program tries up to five total times before moving on. 

Before the server was changed to use encryption, I never saw any exceptions. After the server was changed to use encryption, I started getting occasional exceptions and had to build in a retry loop for each call to a FluentFTP method.

The application program uses FluentFTP's **GetListing()** and **DownloadFile()** methods.

**Connection options:**
```
                    client = New FtpClient(strServerName, strUsername, strPassword) With {
                        .EncryptionMode = FtpEncryptionMode.Explicit,
                        .SslProtocols = Authentication.SslProtocols.Tls12,
                        .DataConnectionConnectTimeout = TIMEOUT,
                        .DataConnectionReadTimeout = TIMEOUT,
                        .ReadTimeout = TIMEOUT,
                        .ConnectTimeout = TIMEOUT,
                        .InternetProtocolVersions = FtpIpVersion.IPv4,
                        .DownloadRateLimit = 0,
                        .RetryAttempts = 2,
                        .TransferChunkSize = 32768,
                        .DataConnectionType = FtpDataConnectionType.PASVEX,
                        .DownloadDataType = FtpDataType.Binary
                    }
```

[log.txt](https://github.com/robinrodricks/FluentFTP/files/3301408/log.txt)

**Comment by robinrodricks** [comment]
Please try this..

https://we.tl/t-SAXE1eEfzz

I think its just an issue of exceptions being thrown in `FtpSocketStream.Dispose`... please try...

**Comment by mikemeinz** [comment]
Using version 25.0.4 - same issue.

In the trace log, line 6895 is where the exception "_Received an unexpected EOF or 0 bytes from the transport stream._" was returned from **GetListing()**.
Immediately following line 6895 is the application program's retry of the **GetListing()** which received a different exception. At the end of the trace log, is the final try which was successful.

[log.txt](https://github.com/robinrodricks/FluentFTP/files/3310303/log.txt)

**Comment by robinrodricks** [comment]
Some fixes released in 26.0.2, please check:

https://www.nuget.org/packages/FluentFTP/26.0.2

**Comment by mikemeinz** [comment]
Tested OK. 

Thank you.
